### ISSUE TITLE
IDA Integration: Work with IDA 7.0 (specifically 7.5+)

### ISSUE THREAD

**Issue by zachriggle** [pwndbg/pwndbg#844]
Currently, the IDA pro integration only works as-is on versions of IDA v7.4 and below.

The most recent v7.5 deprecated the old API by default, and while it can be enabled, the IDAPython script and the `pwndbg.ida` module makes use of the old API.

Since IDA v7.0 was released several years ago, it should be safe to deprecate support entirely for versions 6.x and lower.

**Comment by wtdcode** [comment]
The pwndbg depends on `xmlrpclib` which is removed from Python3. Besides, the current XML server runs in a child thread and IDA API can't be called out of the main thread I believe. Thus, lots of rework should be done...

**Comment by zachriggle** [comment]
Python3 has `xmlrpc.server`, which is effectively the same.

Also, "can't be called" is strong.  Maybe not advised, but I've definitely used it to great success with IDA 7.4.  That said, we do have some custom code to save a timestamped copy of your IDB each time Pwndbg connects, just in case Bad Things (TM) happen.

**Comment by wtdcode** [comment]
Yes, I have tried to do some rework. The real problem is “xmlrpclib.Marshaller”, which don’t have an alternative in Python3.

About my claim about IDA API in multithread environment, I’m sure some API can only be called in the plugin main thread but probably not all of API I guess.

________________________________

**Comment by WeSeekAfterKnowledge** [comment]
I tried to do a bit of work on this, it looks like our mechanism for module registration won't work on the new API.  Trying to see if we can work around this, but even trying to catch the exception (and continue with registration of other functions) it doesn't seem to like it.

```
Python>register_module(idaapi)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<string>", line 3, in register_module
  File "/Applications/IDA Pro 7.5/ida64.app/Contents/MacOS/python/3/ida_idd.py", line 2429, in __getattr__
    ea = self.__name_or_ea(name_or_ea)
  File "/Applications/IDA Pro 7.5/ida64.app/Contents/MacOS/python/3/ida_idd.py", line 2377, in __name_or_ea
    raise ValueError("Undefined function " + name_or_ea)
ValueError: Undefined function __call__
```

**Comment by WeSeekAfterKnowledge** [comment]
Spent a bunch of time getting this to work, unfortunately it looks like @wtdcode was correct about threading issues within IDA 7.x.

```
Exception occurred: context: <Fault 1: "<class 'RuntimeError'>:Function can be called from the main thread only"> (<class 'xmlrpc.client.Fault'>)
```

This happens despite the IDA.py script registering the functions to be called on the main thread using `wrap`:

```py
def wrap(f):
    def wrapper(*a, **kw):
        rv = []
        error = []

        def work():
            try:
                result = f(*a, **kw)
                rv.append(result)
            except Exception as e:
                error.append(e)

        with mutex:
            flags = idaapi.MFF_WRITE
            if f == idc.SetColor:
                flags |= idaapi.MFF_NOWAIT
                rv.append(None)
            idaapi.execute_sync(work, flags)

        if error:
            msg = 'Failed on calling {}.{} with args: {}, kwargs: {}\nException: {}' \
                .format(f.__module__, f.__name__, a, kw, str(error[0]))
            print('[!!!] ERROR:', msg)
            raise error[0]

        return rv[0]

    return wrapper
```

Which uses `idaapi.execute_sync` which I assume would put the function call on the main thread.

**Comment by WeSeekAfterKnowledge** [comment]
Per the jfx-bridge-ida project, using `idaapi.execute_sync` SHOULD work.  This limitation is since IDA 7.2.

https://pypi.org/project/jfx-bridge-ida/

> As of IDA 7.2, all APIs not explicitly marked THREAD_SAFE have to be called from the main thread in IDA. If they aren't, IDA throws a RuntimeError: Function can be called from the main thread only.

It looks like they basically do what we do, but apparently their approach works:

https://github.com/justfoxing/jfx_bridge_ida/blob/2a2a0e6a9822deff2cfa7110208233e28842d2ea/jfx_bridge_ida/server/jfx_bridge_ida_server.py#L41-L76

Compare to:

https://github.com/pwndbg/pwndbg/blob/dev/ida_script.py#L81-L86

Also for my own reference, getting this to work from Docker requires setting the RPC host to `host.docker.internal`

**Comment by WeSeekAfterKnowledge** [comment]
I created a minimal POC that shows that we SHOULD still be able to invoke routines from threads.  I'll have to dig in more as to why this fails when invoked from the XMLRPC server.

The routine this invokes, `idc.get_next_seg`, is one of the ones that fails on my Pull Request #878 but works here.

```py
import idaapi
import idc
import threading

def sync(f):
    def wrapper(*a, **kw):

        def work():
            f(*a, **kw)

        flags = idaapi.MFF_WRITE
        idaapi.execute_sync(work, flags)

    return wrapper

@sync
def here():
    print("Base address is %#x" % idc.get_next_seg(0))

print("[+] Trying from the main thread...")
here()

print("[+] Trying from a different thread")
thread = threading.Thread(target=here)
thread.daemon = True
thread.start()
```

**Comment by WeSeekAfterKnowledge** [comment]
I've extended my example code, and it looks like IDA 7.5 *really* does not like hosting an XMLRPC server.  Simple functions work fine, but invoking anything that calls `idaapi.execute_sync(...)` hangs indefinitely.

```py
import idaapi
import idc
import threading
import time

from xmlrpc.client import ServerProxy
from xmlrpc.server import SimpleXMLRPCServer

def sync(f):
    def wrapper(*a, **kw):

        def work():
            f(*a, **kw)

        flags = idaapi.MFF_WRITE
        idaapi.execute_sync(work, flags)

    return wrapper

def hello():
    print("Hello, world!")

@sync
def here():
    print("Base address is %#x" % idc.get_next_seg(0))

print("[+] Trying from the main thread...")
here()

print("[+] Trying from a different thread")
thread = threading.Thread(target=here)
thread.daemon = True
thread.start()
time.sleep(1)

print("[+] Trying from XMLRPC server")
host = '127.0.0.1'
port = 31337

server = SimpleXMLRPCServer((host, port))
server.register_function(here, "here")
server.register_function(hello, "hello")
server.register_introspection_functions()

thread = threading.Thread(target=server.serve_forever)
thread.daemon = True
thread.start()

client = ServerProxy(f"http://{host}:{port}")

# This works, proving the XMLRPC client and server are communicating
client.hello()

# This hangs forever
# client.here() 
```

**Comment by zachriggle** [comment]
I reached out to Hex-Rays support, and the reason the above ^^ script fails os because `client.here()` is executed on the main thread, and waits for the XMLRPC response.  The execute_sync is waiting for the main thread to be available, which it never will be, because of the main thread waiting on the XMLRPC response.

**Comment by kotee4ko** [comment]
```

#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import print_function

import datetime
import threading
import xmlrpclib
from SimpleXMLRPCServer import SimpleXMLRPCServer
from xml.sax.saxutils import escape

import idaapi
import idautils
import idc

# Wait for any processing to get done
idaapi.auto_wait()

# On Windows with NTFS filesystem a filepath with ':'
# is treated as NTFS ADS (Alternative Data Stream)
# and so saving file with such name fails
dt = datetime.datetime.now().isoformat().replace(':', '-')

# Save the database so nothing gets lost.
if idaapi.IDA_SDK_VERSION >= 700:
    idaapi.save_database(idc.get_idb_path() + '.' + dt, ida_loader.DBFL_KILL)
else:
    idc.SaveBase(idc.GetIdbPath() + '.' + dt)


DEBUG_MARSHALLING = False

def create_marshaller(use_format=None, just_to_str=False):
    assert use_format or just_to_str, 'Either pass format to use or make it converting the value to str.'

    def wrapper(_marshaller, value, appender):
        if use_format:
            marshalled = use_format % value
        elif just_to_str:
            marshalled = '<value><string>%s</string></value>' % escape(str(value))

        if DEBUG_MARSHALLING:
            print("Marshalled: '%s'" % marshalled)

        appender(marshalled)

    return wrapper

xmlrpclib.Marshaller.dispatch[type(0L)] = create_marshaller("<value><i8>%d</i8></value>")
xmlrpclib.Marshaller.dispatch[type(0)] = create_marshaller("<value><i8>%d</i8></value>")
xmlrpclib.Marshaller.dispatch[idaapi.cfuncptr_t] = create_marshaller(just_to_str=True)

host = '127.0.0.1'
port = 8228
#orig_LineA = idc.LineA
orig_LineA = ida_lines.get_extra_cmt


def LineA(*a, **kw):
    v = orig_LineA(*a, **kw)
    if v and v.startswith('\x01\x04; '):
        v = v[4:]
    return v


idc.LineA = LineA

mutex = threading.Condition()


def wrap(f):
    def wrapper(*a, **kw):
        rv = []
        error = []

        def work():
            try:
                result = f(*a, **kw)
                rv.append(result)
            except Exception as e:
                error.append(e)

        with mutex:
            flags = idaapi.MFF_WRITE
            if f == idc.SetColor:
                flags |= idaapi.MFF_NOWAIT
                rv.append(None)
            idaapi.execute_sync(work, flags)

        if error:
            msg = 'Failed on calling {}.{} with args: {}, kwargs: {}\nException: {}' \
                .format(f.__module__, f.__name__, a, kw, str(error[0]))
            print('[!!!] ERROR:', msg)
            raise error[0]

        return rv[0]

    return wrapper


def register_module(module):
    for name, function in module.__dict__.items():
        if hasattr(function, '__call__'):
            server.register_function(wrap(function), name)


def decompile(addr):
    """
    Function that overwrites `idaapi.decompile` for xmlrpc so that instead
    of throwing an exception on `idaapi.DecompilationFailure` it just returns `None`.
    (so that we don't have to parse xmlrpc Fault's exception string on pwndbg side
    as it differs between IDA versions).
    """
    try:
        return idaapi.decompile(addr)
    except idaapi.DecompilationFailure:
        return None

def get_decompile_coord_by_ea(cfunc, addr):
    if idaapi.IDA_SDK_VERSION >= 720:
        item = cfunc.body.find_closest_addr(addr)
        y_holder = idaapi.int_pointer()
        if not cfunc.find_item_coords(item, None, y_holder):
            return None
        y = y_holder.value()
    else:
        lnmap = {}
        for i, line in enumerate(cfunc.pseudocode):
            phead = idaapi.ctree_item_t()
            pitem = idaapi.ctree_item_t()
            ptail = idaapi.ctree_item_t()
            ret = cfunc.get_line_item(line.line, 0, True, phead, pitem, ptail)
            if ret and pitem.it:
                lnmap[pitem.it.ea] = i
        y = None
        closest_ea = BADADDR
        for ea,line in lnmap.items():
            if closest_ea == BADADDR or abs(closest_ea - addr) > abs(ea - addr):
                closest_ea = ea
                y = lnmap[ea]

    return y


def decompile_context(addr, context_lines):
    cfunc = decompile(addr)
    if cfunc is None:
        return None
    y = get_decompile_coord_by_ea(cfunc, addr)
    if y is None:
        return cfunc
    lines = cfunc.get_pseudocode()
    retlines = []
    for lnnum in range(max(0, y - context_lines), min(len(lines), y + context_lines)):
        retlines.append(idaapi.tag_remove(lines[lnnum].line))
        if lnnum == y:
            retlines[-1] = '>' + retlines[-1][1:]
    return '\n'.join(retlines)


def versions():
    """Returns IDA & Python versions"""
    import sys
    return {
        'python': sys.version,
        'ida': idaapi.get_kernel_version(),
        'hexrays': idaapi.get_hexrays_version() if idaapi.init_hexrays_plugin() else None
    }


server = SimpleXMLRPCServer((host, port), logRequests=True, allow_none=True)
register_module(idc)
register_module(idautils)
register_module(idaapi)
server.register_function(lambda a: eval(a, globals(), locals()), 'eval')
server.register_function(wrap(decompile)) # overwrites idaapi/ida_hexrays.decompile
server.register_function(wrap(decompile_context), 'decompile_context')  # support context decompile
server.register_function(versions)
server.register_introspection_functions()

print('IDA Pro xmlrpc hosted on http://%s:%s' % (host, port))
print('Call `shutdown()` to shutdown the IDA Pro xmlrpc server.')

thread = threading.Thread(target=server.serve_forever)
thread.daemon = True
thread.start()


def shutdown():
    global server
    global thread
    server.shutdown()
    server.server_close()
    del server
    del thread


```

this changes according to https://www.hex-rays.com/products/ida/support/ida74_idapython_no_bc695_porting_guide.shtml
allow me to get next result:

```
Python> (2.7, windows)
Flushing buffers, please wait...ok
IDA Pro xmlrpc hosted on http://127.0.0.1:8228
Call `shutdown()` to shutdown the IDA Pro xmlrpc server.
```

**Comment by kotee4ko** [comment]
On IDA Pro 7.5 *

**Comment by kotee4ko** [comment]
xmlrpc.client.Fault: <Fault 1: "<type 'exceptions.AttributeError'>:'module' object has no attribute 'SetColor'">

now, I'm see this on client-side (GDB).

**Comment by kotee4ko** [comment]
It seems that in `pwndbg/ida.py` a lot's of old-style classes and methods. I try to fix it, now I get this error:
```Exception occurred: Error: <Fault 1: "<type 'exceptions.TypeError'>:cannot marshal <type 'SwigPyObject'> objects"> (<class 'xmlrpc.client.Fault'>)```

It's hard to find entry point of this. May be somebody know?

**Comment by kotee4ko** [comment]
idk, but it's seems that integrate PWNGDB's `nextcall` `nextret` `vis` to gef - is much simple, that integrate IDA and BN to  PWNGDB

**Comment by maskelihileci** [comment]
> Hex-Rays desteğine ulaştım ve yukarıdaki ^^ betiğinin işletim sisteminde başarısız olmasının nedeni `client.here()`ana iş parçacığında yürütülüyor ve XMLRPC yanıtını bekliyor. Execute_sync, XMLRPC yanıtını bekleyen ana iş parçacığı nedeniyle, hiçbir zaman olmayacak olan ana iş parçacığının kullanılabilir olmasını bekliyor.

Ask them why they still use a single core in 2022
I to Waiting for hours for a process
If it hadn't had good support for the plugin, I would have switched to other software.

**Comment by mahaloz** [comment]
Hi @maskelihileci (and devs of pwndbg). A while back I looked into doing this, but it actually meant a major refactor of the code. You might be more interested in using this plugin I made for pwndbg/gef/gdb which is updated for 7.0 API and a few other nifty features:
https://github.com/mahaloz/decomp2dbg

At the very least, if you are dead set on doing this and the devs want it in their master branch you can use my code which has already done all this work of finding the right API and wrapping it for Main Thread: 
https://github.com/mahaloz/decomp2dbg/blob/main/decompilers/d2d_ida/d2d_ida/server.py

However, plz cite in the commits and code where you got it from ;).

**Comment by disconnect3d** [comment]
@mahaloz I'd be fine & happy with using your code (and ofc giving you credits for that!) but I haven't used IDA integration for a long time since I don't have an official license and the free one I am using now is missing Python scripting capabilities.

If there's a chance you or somebody else can contribute that, I'd love merging such PR.

**Comment by disconnect3d** [comment]
This is old & stale. We have some integrationprovider abstraction this can be implemented with but i dont think it is right now. Closing
